include/exadg/operators/finite_element.h

```diff
@@ -0,0 +1,128 @@
+/*  ______________________________________________________________________
+ *
+ *  ExaDG - High-Order Discontinuous Galerkin for the Exa-Scale
+ *
+ *  Copyright (C) 2023 by the ExaDG authors
+ *
+ *  This program is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 3 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *  ______________________________________________________________________
+ */
+
+#ifndef INCLUDE_EXADG_OPERATORS_FINITE_ELEMENT_H_
+#define INCLUDE_EXADG_OPERATORS_FINITE_ELEMENT_H_
+
+// deal.II
+#include <deal.II/fe/fe_dgq.h>
+#include <deal.II/fe/fe_q.h>
+#include <deal.II/fe/fe_simplex_p.h>
+#include <deal.II/fe/fe_system.h>
+
+// ExaDG
+#include <exadg/grid/grid_data.h>
+#include <exadg/utilities/exceptions.h>
+
+namespace ExaDG
+{
+/**
+ * Creates and returns a shared pointer to dealii::FiniteElement<dim> depending on the element type,
+ * the type of function space, the number of components of the solution field, and the polynomial
+ * degree of the shape functions. Regarding the function space, the code assumes an H^1 conforming
+ * space (Continuous Galerkin) if is_dg (= L^2-conforming space, Discontinuous Galerkin) is false.
+ */
+template<int dim>
+std::shared_ptr<dealii::FiniteElement<dim>>
+create_finite_element(ElementType const & element_type,
+                      bool const          is_dg,
+                      unsigned int const  n_components,
+                      unsigned int const  degree)
+{
+  std::shared_ptr<dealii::FiniteElement<dim>> fe;
+
+  if(n_components == 1)
+  {
+    if(is_dg)
+    {
+      if(element_type == ElementType::Hypercube)
+      {
+        fe = std::make_shared<dealii::FE_DGQ<dim>>(degree);
```

Comment:
I think that in the longer run, we would probably want to increase the flexibility here. For example, in the future we will probably want to use `FE_DGQHermite` in some circumstances, and `FE_DGQArbitraryNodes<dim>(QGauss<dim>(degree + 1))` in some others because they might be beneficial for certain setups. That is a longer-term objective that cannot be resolved in this PR, but I would like to plan for how we would support it. Maybe the comment above regarding `Conformity` is then not the right choice, but rather an enum encoding all types. The important part is that we should have a name that does not care about whether we have `Hypercube`, `Simplex` or (at a later stage), `Wedge` and `Pyramid` elements for supporting mixed geometries. The advantage would then be a more flexible choice in the space at the place where we call `create_finite_element` by this enum, and yet be flexible in our interfaces. What do you think?

Reply:
I think one would need several enums. One for CG/DG/Hdiv/etc., one for GL/GLL/Hermite/etc., one for the element type. Not all combinations of these enums will be supported, which needs to be asserted in my opinion.